jujugui/static/gui/src/app/store/env/bakery.js

```diff
@@ -35,302 +35,313 @@ YUI.add('juju-env-bakery', function(Y) {
+
+     @method _requestHandler
+     @param {Function} successCallback Called when the api request completes
+            successfully.
+     @param {Function} failureCallback Called when the api request fails
+            with a response of >= 400.
+     @param {Object} response The XHR response object.
+     @return {undefined} Nothing.
+     */
+    _requestHandler: function (successCallback, failureCallback, response) {
+      var target = response.target;
+      if (target.status >= 400) {
+        failureCallback(response);
+        return;
+      }
+      successCallback(response);
+    },
+
+    /**
+     Used to resend the original request without any interaction this time.
+
+     @method _sendOriginalRequest
+     @param {String} The path to make the api request to.
+     @param {Function} successCallback Called when the api request completes
+            successfully.
+     @param {Function} failureCallback Called when the api request fails
+            with a response of >= 400 (except 401/407).
+     @return {undefined} Nothing.
+     */
+    _sendOriginalRequest: function(path, successCallback, failureCallback) {
+      var macaroons = this._getMacaroon();
+      var headers = {'Bakery-Protocol-Version': 1};
+      if (macaroons !== null) {
+        headers['Macaroons'] = macaroons;
+      }
+      this.webhandler.sendGetRequest(
+        path, headers, null, null, false, null,
+        this._requestHandler.bind(
+          this, successCallback, failureCallback
+        )
+      );
+    },
+
+    /**
+     Authenticate by discharging the macaroon and
+     then set the cookie by calling the authCookiePath provided.
+
+     @method authenticate
+     @param {Macaroon} The macaroon to be discharged.
+     @param {Function} The request to be sent again in case of
+            successful authentication.
+     @param {Function} The callback failure in case of wrong authentication.
+     @return {undefined} Nothing.
+     */
+    _authenticate: function (m, requestFunction, failureCallback) {
+      try {
+        macaroon.discharge(
+          macaroon.import(m),
+          this._obtainThirdPartyDischarge.bind(this),
+          this._processDischarges.bind(
             this,
-            this._setMacaroonsCookie.bind(this, requestFunction, jsonMacaroon),
+            requestFunction,
             failureCallback
-          )
+          ),
+          failureCallback
         );
-      },
-
-      /**
-       Process successful discharge by setting Macaroons Cookie
-       and invoke the original request
-
-       @method _setMacaroonsCookie
-       @param {Function} The path where to send put request
-              to set the cookie back
-       @param {Object} an exported Macaroon
-       */
-      _setMacaroonsCookie: function (originalRequest, jsonMacaroon) {
-        var prefix = this.macaroonName + '=';
-        document.cookie = prefix + btoa(JSON.stringify(jsonMacaroon));
-        window.addEventListener('unload', function() {
```

Comment:
If we're losing this as part of the change from  `_setMacaroonsCookie` to `_processDischarges`, what's managing removing the cookie on logout?